Entrar/dejar un rol - Fixed darkaqua/darkbot_old#69

TheCreativeGod opened the pull request:
Permite a un usuario entrar o dejar ciertos roles
Uso: `!rol join/leave [name]`
Ejemplo: `!rol join [coder]`

TheCreativeGod committed:
Entrar/dejar un rol

Permite a un usuario entrar o dejar un rol con !rol join/leave [nombre]

mcmacker4 commented on bot/commands/cm_rol.js:
```diff
@@ -0,0 +1,39 @@
+  let roleName = args[2].replace(/_/g, " ");
+  let rol = (() => {
+      return args[2] ? message.guild.roles.findKey("name", roleName) : "";
+  })();
```
Aqui no necesitas un closure, no hay ninguna variable para esconder.
```js
let rol = args[2] ? message.guild.roles.findKey("name", roleName) : "";
```

TheCreativeGod replied:
Necesito `roleName` para probar que esté en la lista, `rol` me da la ID o en ese caso tendría que poner ID's en lugar de nombres en la lista de roles(?).
Si hago
```JS
let rol = args[2] ? message.guild.roles.findKey("name", roleName) : "";
```
si el rol tiene espacios habrá u problema por ahí... por ahora no habría problema, pero si en un futuro deciden crear un rol con espacios si los habrá.(?)

TheCreativeGod replied:
Wait... what? me dió una embolía... no sé que leí la 1ra vez...

mcmacker4 replied:
No he dicho nada de roleName, lo que has hecho aquí de `(() => { return cosa; })()` es un closure, se suele usar para esconder variables, pero ahí no es necesario porque no hay ninguna variable a esconder.

Sobre lo de los espacios, realmente no es necesario escribir el rol con barras bajas porque el argumento de role es el ultimo. En un comando `!role join un rol con espacios` sabes que el primer argumento es `role` el segundo sera `join` o `leave` y a partir de ahí, podrías juntar el resto de argumentos (`un rol con espacios`) en uno solo usando espacios como separador, et voilà, tienes un bonito rol con espacios.

mcmacker4 commented on bot/commands/cm_rol.js:
```diff
@@ -0,0 +1,39 @@
+}
+message.delete();
+}
+}
```
Que ha pasado con las tabulaciones?? :fearful:
La mayoria de editores de codigo tienen un atajo para arreglar el formato del texto automaticamente. Usalo pls.

mcmacker4 commented on bot/commands/cm_rol.js:
```diff
@@ -0,0 +1,39 @@
+//Roles asignados de manera especial y de Administración
+const special = ["[admin]", "[sub-admin]", "[project]", "[bot]", "[darkbot coder]", "tercer componente de Daft Punk", "[salt]", "[helper]"];
```
Maybe añadir `[salt]` aqui? Se le asigna a los jugadores del LoL automaticamente.

mcmacker4 replied:
Y ahora que lo pienso, igual esto debería ir al `config.json` para acceder desde `global.config`, y no deberían ser los protegidos, sino los que se pueden usar.

pagoru replied:
¿Mejor como archvio de constantes dentro del proyecto, junto a la id de las salas?

mcmacker4 reviewed (changes requested):
En cuanto arregles lo que te he dicho, lo pruebo en mi servidor y si funciona hago el merge.

pagoru committed:
Merge branch 'develop' into tcg-rol

TheCreativeGod committed:
Merge branch 'develop' into tcg-rol

TheCreativeGod committed:
#34 , indentado, arreglos varios.

Arreglo al #22 e implementación del #34

TheCreativeGod commented on bot/commands/cm_rol.js:
```diff
@@ -0,0 +1,32 @@
+		if (args[1] && args[2]) {
+			let roleName = args.splice(2).join(' ');
+			let rol = roleName ? message.guild.roles.findKey("name", roleName) : "";
+			if (rol && global.config.roles.includes(roleName)) {
```
Se necesita agregar
```JSON
"roles": ["lista", "de", "roles"]
```
A `config.json`

pagoru replied:
Cambia el `README.md` !!

TheCreativeGod replied:
Hecho borquecho.

TheCreativeGod committed:
Adiós data.json A.K.A: #34

Ya no se usará el archivo data.json para leer atributos, el archivo se
eliminó.

TheCreativeGod committed:
Update README.md

mcmacker4 commented on bot/commands/cm_rol.js:
```diff
@@ -0,0 +1,32 @@
+				//busca que el argumento sea "leave" o "join"
+				if (args[1] == "join") {
+					message.member.addRole(rol);
+					message.author.sendMessage(`Te has unido a \`${roleName}\``);
```
`member.addRole(rol)` devuelve una Promise que podria lanzar un error. Si salta error saltara el maldito `Unhandled Promise Rejection Error` que no te dice donde ha saltado el error. A parte, si salta el error aun asi se enviara el mensaje de `Te has unido` cuando realmente no ha ocurrido. Yo lo cambiaria por esto (y en el leave tambien):

```js
message.member.addRole(rol).then(member =>
    message.author.sendMessage(`Te has unido a \`${roleName}\``)
).catch(err => {
    message.author.sendMessage("Se ha producido un error al unirte al rol.");
    console.log(err.message);
});
```

mcmacker4 reviewed (changes requested):
Solo un cambio mas y te dejo en paz (sorry).

mcmacker4 committed:
Unhandled Promise Rejection

Pequeños cambios con los Promises.

pagoru committed:
Reestructurando el código

pagoru committed:
Merge branch 'develop' into tcg-rol

pagoru commented:
Fixed darkaqua/darkbot_old#69
